src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -2288,6 +2289,46 @@ SyncNodeMetadataToNodes(void)
 }
 
 
+/*
+ * SyncNodeMetadataToNodes recreates the node metadata snapshot in all the
+ * metadata workers.
+ *
+ * This function runs within a coordinated transaction since updates on
+ * the pg_dist_node metadata must be rollbacked if anything
+ * goes wrong.
+ */
+void
+SyncNodeMetadataToNodes(void)
+{
+	EnsureCoordinator();
+
+	/*
+	 * Request a RowExclusiveLock so we don't run concurrently with other
+	 * functions updating pg_dist_node, but allow concurrency with functions
+	 * which are just reading from pg_dist_node.
+	 */
+	if (!ConditionalLockRelationOid(DistNodeRelationId(), RowExclusiveLock))
+	{
+		ereport(ERROR, (errmsg("cannot sync metadata because a concurrent "
+							   "metadata syncing operation is in progress")));
+	}
+
+	List *workerList = ActivePrimaryNonCoordinatorNodeList(NoLock);
+	WorkerNode *workerNode = NULL;
+	foreach_ptr(workerNode, workerList)
+	{
+		if (workerNode->hasMetadata)
+		{
+			SetWorkerColumnLocalOnly(workerNode, Anum_pg_dist_node_metadatasynced,
+									 BoolGetDatum(true));
+
+			bool raiseOnError = true;
+			SyncNodeMetadataSnapshotToNode(workerNode, raiseOnError);
```

Comment:
Hmm, do we additionally need to wait all the active transactions on all nodes to finish? Because otherwise as the COMMITs will happen in different times on the nodes, we might still have some the data divergence. 

The window is small, but still..

Reply:
Can you explain the scenario a little more. Presumably replacing all the metadata blocks everything.

Reply:
1) User does `citus_disable_node(`worker-1`, force:=true, sync:=true)`;
2) On worker-2, PREPARE TRANSACTION 'xxxx'; executed
3) On worker-3, PREPARE TRANSACTION 'xxxx'; executed
4) On worker-2, COMMIT PREPARED 'xxxx'; executed
5) On worker-2, UPDATE TO REFERENCE table acquired the remote lock locally (as it is the first node now)
6) On worker-3, UPDATE TO REFERENCE table acquired the remote lock on `worker-1` as (3) has not been committed yet
7) Updates on (6) and (7) are serialized on different worker node, hence risk of divergence exists


One thing to note that, node metadata sync sends `DELETE FROM pg_dist_node`. If it was `TRUNCATE pg_dist_node`, operation (6) would have been blocked on pg_dist_node?

Reply:
> **One thing to note that, node metadata sync sends DELETE FROM pg_dist_node. If it was TRUNCATE pg_dist_node, operation (6) would have been blocked on pg_dist_node?

As far as I remember, we changed this to DELETE to prevent certain distributed deadlocks on concurrent metadata sync vs metadata change operations. Maybe we should check if we can make it `TRUNCATE` again

Reply:
I pushed a commit, which is a little aggressive by actually stopping `all` the traffic in the cluster until the commit. As I noted in the comments, we could come-up with something smarter that only blocks modifications to the replicated tables, which does not exists (as far as I can see).

- Are you aware of such mechanism? Locking `pg_dist_transaction` doesn't seem to work fine because that happens after the remote locks are acquired.
- Do you think is this OK for now?
- As far as I can see, the more proper approach could be to run a variation of `StartLockAcquireHelperBackgroundWorker` on *ALL* worker nodes concurrently. We might need this anyway when we support citus 11 + pg_auto_failover. For now, that seemed too much effort for this case.